polkadot/node/collation-generation/src/lib.rs

```diff
@@ -596,25 +532,47 @@ async fn construct_and_distribute_receipt(
 		hrmp_watermark: collation.hrmp_watermark,
 	};
 
-	let ccr = CandidateReceipt {
-		commitments_hash: commitments.hash(),
-		descriptor: CandidateDescriptor {
-			signature: key.sign(&signature_payload),
-			para_id,
-			relay_parent,
-			collator: key.public(),
-			persisted_validation_data_hash,
-			pov_hash,
-			erasure_root,
-			para_head: commitments.head_data.hash(),
-			validation_code_hash,
+	let receipt = if v2_receipts {
+		let ccr = CommittedCandidateReceiptV2 {
+			descriptor: CandidateDescriptorV2::new(
+				para_id,
+				relay_parent,
+				core_index,
+				session_index,
+				persisted_validation_data_hash,
+				pov_hash,
+				erasure_root,
+				commitments.head_data.hash(),
+				validation_code_hash,
+			),
+			commitments,
+		};
+
+		ccr.check_core_index(&transposed_claim_queue)
+			.map_err(Error::CandidateReceiptCheck)?;
+
+		ccr.to_plain()
+	} else {
+		CandidateReceipt {
+			commitments_hash: commitments.hash(),
+			descriptor: CandidateDescriptor {
+				signature: key.sign(&signature_payload),
+				para_id,
+				relay_parent,
+				collator: key.public(),
+				persisted_validation_data_hash,
+				pov_hash,
+				erasure_root,
+				para_head: commitments.head_data.hash(),
```

Comment:
Perhaps it is wise to add a warning if the commitments commit to a core while the v2 receipts are disabled.